src/MANYBODY/pair_rebo.h

```diff
@@ -28,6 +28,7 @@ class PairREBO : public PairAIREBO {
  public:
   PairREBO(class LAMMPS *);
   void settings(int, char **);
+  void spline_init();
```

Comment:
So, for starters, the corresponding method in `pair_airebo.h` should be marked `virtual`.  Alas, that still does not fix the issue.

Reply:
No, I mean, if you don't declare the superclass function as `virtual`, then it can't be overridden by the one on the derived class.

> however, since the method of the derived class is run after the base class, the final effect should be comparable.

I don't know what you're saying here.  If you add simple `printf` debugging statements, you will see that, without the `virtual` keyword, `PairREBO::spline_init` is never run.  Only the one on `PairAIREBO` is.

Reply:
Or maybe you meant to add a call to it from the derived class constructor?

Reply:
@ExpHP thanks for the clarification. The issue is more complex. C++ scoping is sometimes rather confusing. Under normal circumstances, if you derive a class and have no polymorphism (i.e. virtual methods), the method of the same name in the derived class is called as well as the method in the base class and after the base class method. However, in this case, the `spline_init()` method is called from a polymorph (virtual) method and thus the scoping rules are different. In LAMMPS the whole thing gets even more complicated, by having multiple "pathways" to derive the REBO class, due to the USER-OMP package. I've made a few tests and noticed, that there are bugs (missing statements to set flags) in checking for the potential file consistency there. I'm doing a little more cleaning up and update to the latest master branch to have some other issues sorted out, too.

Reply:
>  Under normal circumstances, if you derive a class and have no polymorphism (i.e. virtual methods), the method of the same name in the derived class is called as well as the method in the base class and after the base class method.

This sounds extremely bizarre to me and I've never heard of anything like this except for constructors and destructors.  It is not clear to me how this works for methods with non-`void` return types, or methods with movable arguments (of type `&&T`).

The rules as I understand it are simple if you think about it in terms of what fully-compiled, optimized code is capable of doing.  Any method call always calls a single method.  Normally, the compiler knows exactly what method is going to be called, and may even inline it (demonstrating that the compiler does not care about the possibility of derived class methods having the same name).  Marking a method as `virtual` on the other hand makes it opaque to the compiler.  Every call to it to must be dynamically looked up through a vtable, allowing it to be overridden.

`pair_airebo.o` is compiled without any knowledge of the `PairREBO` class, so there is no way for any code there to call methods from the `PairREBO` class except indirectly through a vtable lookup.